frappe/public/js/frappe/form/log_call.js

```diff
@@ -0,0 +1,177 @@
+            "communication_type": ["in", ["Comment", "Communication"]]
+          }
+        },
+        callback: function(r) {
+          promptInfo.push({fieldtype: "Section Break"})
+          if(r.message != undefined) {
+            r.message.forEach(function(messageData) {
+              var content = ""
+              if(messageData.Communication_Type == "Comment") {
+                if(messageData.Content != null && messageData.Subject != null && messageData.Content < messageData.Subject) {
+                  content = messageData.Subject + '\n' + messageData.Content
+                } else if(messageData.Content != null) {
+                  content = messageData.Content
+                } else if (foo.Subject != null) {
+                  content = messageData.Subject
+                }
+                newMessage = {label: "<i class='octicon octicon-comment-discussion icon-fixed-width'></i>Comment - " + messageData.Communication_Date, fieldname: "message",
+                              fieldtype: "Small Text", default: content, read_only: "1"}
+              } else if (foo.Communication_Type == "Communication") {
+                content = messageData.Subject + '\n' + messageData.Content
+                newMessage = {label: "<i class='octicon octicon-device-mobile icon-fixed-width'></i> Communication", fieldname: "message",
+                              fieldtype: "Small Text", default: content, read_only: "1"}
+              }
+              var state = true
+              promptInfo.forEach(function(bar) {
+                if(JSON.stringify(bar) == JSON.stringify(newMessage)) {
+                  state = false
+                }
+              })
+              if(state) {
+                promptInfo.push(newMessage);
+                promptInfo.push({fieldtype: "Section Break"});
+              }''
+            })
+          }
+          var communication = frappe.prompt(promptInfo,
+            function(data){
+              frm.doc.contact_date = data.next_contact_date;
+              frm.doc.contact_by = frappe.user.name;
+              frappe.call({
+                method: "erpnext_controllers.erpnext_controllers.methods.log_call.update_doc",
+                args: {
+                  "doc": frm.doc
+                },
+                callback: function(r) {
+                  frappe.call({
+                    method: "erpnext_controllers.erpnext_controllers.methods.log_call.make_communication",
+                    args: {
+                      "doc": frm.doc,
+                      "communication": data
+                    },
+                    callback: function(r){
+                      cur_frm.refresh();
+                    }
+                  })
+                }
+              })
+            });
+          }
+        });
+    });
+});
+
+frappe.ui.form.on("Opportunity", "refresh", function(frm) {
+    var promptInfo = [
+        {label: "Contact", fieldtype: "Select", options: frm.doc.customer_name,
```

Comment:
The forms have a different fieldname for 'contact', so the prompts are slightly different. Not all documents have a 'next contact date' field as well (but both of these do).

One of the things to review is what information is important to display to the user when using this feature - and determine a good way to generalize that across different documents. It's possible to describe a generic form (with content, subject, etc) and then push things onto that from each document, but that also gets repetitive.

What do you think is a good solution to this?